contracts/multisig/contracts/src/AssetProxyOwner.sol

```diff
@@ -16,93 +16,205 @@
+    function registerFunctionCall(
+        bool hasCustomTimeLock,
+        bytes4 functionSelector,
+        address destination,
+        uint128 newSecondsTimeLocked
+    )
+        external
         onlyWallet
-        notNull(assetProxyContract)
     {
-        isAssetProxyRegistered[assetProxyContract] = isRegistered;
-        emit AssetProxyRegistration(assetProxyContract, isRegistered);
+        _registerFunctionCall(
+            hasCustomTimeLock,
+            functionSelector,
+            destination,
+            newSecondsTimeLocked
+        );
     }
 
-    /// @dev Allows execution of `removeAuthorizedAddressAtIndex` without time lock.
+    /// @dev Allows anyone to execute a confirmed transaction.
+    ///      Transactions *must* encode the values with the signature "bytes[] data, address[] destinations, uint256[] values"
+    ///      The `destination` and `value` fields of the transaction in storage are ignored.
+    ///      All function calls must be successful or the entire call will revert.
     /// @param transactionId Transaction ID.
-    function executeRemoveAuthorizedAddressAtIndex(uint256 transactionId)
+    function executeTransaction(uint256 transactionId)
         public
         notExecuted(transactionId)
         fullyConfirmed(transactionId)
-        validRemoveAuthorizedAddressAtIndexTx(transactionId)
     {
-        Transaction storage txn = transactions[transactionId];
-        txn.executed = true;
-        if (_externalCall(txn.destination, txn.value, txn.data.length, txn.data)) {
-            emit Execution(transactionId);
+        Transaction storage transaction = transactions[transactionId];
+        transaction.executed = true;
+
+        // Decode batch transaction data from transaction.data
+        // `destination` and `value` fields of transaction are ignored
+        // Note that `destination` must be non-0, or the transaction cannot be submitted
+        // solhint-disable
+        (
+            bytes[] memory data,
+            address[] memory destinations,
+            uint256[] memory values
+        ) = abi.decode(
+            transaction.data,
+            (bytes[], address[], uint256[])
+        );
+        // solhint-enable
+
+        // Ensure lengths of array properties are equal
+        uint256 length = data.length;
+        require(
+            length == destinations.length && length == values.length,
+            "EQUAL_LENGTHS_REQUIRED"
+        );
+
+        uint256 transactionConfirmationTime = confirmationTimes[transactionId];
+        for (uint i = 0; i != length; i++) {
+            // Ensure that each function call is past its timelock
+            _assertValidFunctionCall(
+                transactionConfirmationTime,
+                data[i],
+                destinations[i]
+            );
+            // Call each function
+            // solhint-disable-next-line avoid-call-value
+            (bool didSucceed,) = destinations[i].call.value(values[i])(data[i]);
```

Comment:
Could there be a benefit to propagating the return value / revert reason?

Reply:
I'm fine with it either way (see above comment). This should be used pretty infrequently in practice as well.

Reply:
That makes sense. Could be good for future-proofing, but not critical.